subsys/net/ip/tcp.c

```diff
@@ -1117,10 +1117,15 @@ static int tcp_send_queued_data(struct tcp *conn)
 
 	/* If we have out-of-bufs case, then do not start retransmit timer
 	 * yet. The socket layer will catch this and resend data if needed.
+	 * Only perform this when it is just the newly added packet,
```

Comment:
It's not clear to me why is ths a problem when `tcp_send_queued_data()` is called from within `tcp_in()` - this timer should've been canceled already:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp.c#L2033

And I don't see where could it be rescheduled again inbetween.

Reply:
Hmmm.... maybe you are right

Reply:
Thinking about it a little longer, there is still an option when data is the queue and a new packet is added from net_tcp_queue_data. In that case the failing new packet could potentially kill the retransmit timer, where it should be still running.

Reply:
I can simplify this commitment by moving the cancel statement into net_tcp_queue_data. The send timer can be cancelled when the send_data_total reaches 0 after an -ENOBUFS error.

Reply:
The simplification has been implemented in https://github.com/zephyrproject-rtos/zephyr/pull/45890/commits/c6b1973d5567b7d3f7f96793ac5fb54c0c3213ab. It required some magic rebasing to get the other 2 commits also properly in. The end result is a much cleaner tcp_send_queued_data function.

@rlubos Currently the send_timer can be triggered if there is no data in the transmission queue and the transmission window is 0. Do you think that is correct? Otherwise we have to remove 
'if (tcp_window_full(conn)) {
		subscribe = true;
}'
from tcp_send_queued_data

Reply:
I don't think there's a point scheduling the retransmission timer with zero-length window, I faced similar issue here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp.c#L2285
IMO in such case we need to either wait for an ACK from the peer side (w/o scheduling retransmission), or rely on the ZWP timer and send ZWP when neeeded.